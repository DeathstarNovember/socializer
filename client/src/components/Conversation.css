.chat-layout {
  flex: 1;
  height: 100%;
}

.chat-content {
  flex: 1;
}

@media (max-device-width: 768px) {
  .chat-layout.pb-4 {
    padding-bottom: 0px !important;
  }

  .chat-content {
    padding-left: 15px;
    padding-right: 15px;
  }
}
