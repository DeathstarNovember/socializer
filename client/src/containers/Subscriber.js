import { useEffect } from "react";

const Subscriber = ({ subscribeToNew, children }) => {
  useEffect(() => {
    subscribeToNew();
  }, [subscribeToNew]);

  return children;
};

export default Subscriber;
