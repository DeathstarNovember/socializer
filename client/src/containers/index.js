export { default as StateProvider } from "./StateProvider";
export { default as Subscriber } from "./Subscriber";
