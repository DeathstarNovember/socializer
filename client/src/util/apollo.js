import ApolloClient from "apollo-client";
import { InMemoryCache } from "apollo-cache-inmemory";
import * as AbsintheSocket from "@absinthe/socket";
import { createAbsintheSocketLink } from "@absinthe/socket-apollo-link";
import { Socket as PhoenixSocket, LongPoll } from "phoenix";
import { createHttpLink } from "apollo-link-http";
import { hasSubscription } from "@jumpn/utils-graphql";
import { split } from "apollo-link";
import { setContext } from "apollo-link-context";
import Cookies from "js-cookie";

const HTTP_URI =
  process.env.NODE_ENV === "production"
    ? "https://brisk-hospitable-indianelephant.gigalixirapp.com"
    : "http://localhost:4000";

const WS_URI =
  process.env.NODE_ENV === "production"
    ? "wss://brisk-hospitable-indianelephant.gigalixirapp.com/socket"
    : "ws://localhost:4000/socket";

function createPhoenixSocket() {
  // Socket with full fallback to LongPoll
  // via https://elixirforum.com/t/fall-back-to-longpoll-when-websocket-fails/23894
  const socket = new PhoenixSocket(WS_URI, {
    params: () => {
      if (Cookies.get("token")) {
        return { token: Cookies.get("token") };
      } else {
        return {};
      }
    },
  });
  socket.onError(() => {
    if (navigator.onLine) {
      const isWebSocket = socket.transport === window.WebSocket;
      console.error(
        `Error connecting using ${isWebSocket ? "WebSocket" : "LongPoll"}`,
      );
      if (isWebSocket) socket.transport = LongPoll;
      else if (window.WebSocket) socket.transport = window.WebSocket;
    }
  });
  return socket;
}

export const createClient = ({ ssr, req, fetch, tokenCookie } = {}) => {
  let link = createHttpLink({ uri: HTTP_URI, fetch });
  let absintheSocket;

  if (!ssr) {
    absintheSocket = AbsintheSocket.create(createPhoenixSocket());
    const socketLink = createAbsintheSocketLink(absintheSocket);

    link = split(
      (operation) => hasSubscription(operation.query),
      socketLink,
      link,
    );
  }

  const authLink = setContext((_, { headers }) => {
    // Get the authentication token from the cookie if it exists.
    const token = tokenCookie || Cookies.get("token");

    // Return the headers to the context so httpLink can read them.
    return {
      headers: {
        ...headers,
        authorization: token ? `Bearer ${token}` : "",
      },
    };
  });

  const cache = new InMemoryCache();
  if (!ssr) {
    cache.restore(window.__APOLLO_STATE__);
  }

  const client = new ApolloClient({
    connectToDevTools: !ssr,
    ssrMode: ssr,
    link: authLink.concat(link),
    cache,
  });
  return { client, absintheSocket };
};

export const refreshSocket = (socket) => {
  // Close the connection to force a reconnection with the
  // new token parameter.
  socket.phoenixSocket.conn && socket.phoenixSocket.conn.close();
};
