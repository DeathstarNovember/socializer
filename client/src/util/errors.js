import { camelizeKeys } from "humps";

export const errorHash = ({ graphQLErrors } = {}) => {
  if (!Array.isArray(graphQLErrors)) return {};
  return camelizeKeys(
    graphQLErrors.reduce(
      (hash, error) => Object.assign(hash, { [error.field]: error.message }),
      {},
    ),
  );
};
