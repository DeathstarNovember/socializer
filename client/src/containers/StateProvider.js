import React, { useEffect, useState } from "react";
import { useApolloClient } from "@apollo/react-hooks";
import Cookies from "js-cookie";
import { refreshSocket } from "util/apollo";
import { AuthContext, ChatContext } from "util/context";

const StateProvider = ({ initialToken, initialUserId, socket, children }) => {
  const client = useApolloClient();
  const [token, setToken] = useState(initialToken || Cookies.get("token"));
  const [userId, setUserId] = useState(initialUserId || Cookies.get("userId"));
  const [chatState, setChatState] = useState("default");

  // If the token changed (i.e. the user logged in
  // or out), clear the Apollo store and refresh the
  // websocket connection.
  useEffect(() => {
    if (!token) client.clearStore();
    if (socket) refreshSocket(socket);
  }, [client, socket, token]);

  const setAuth = (data) => {
    if (data) {
      const { id, token } = data;
      Cookies.set("token", token);
      Cookies.set("userId", id);
      setToken(token);
      setUserId(id);
    } else {
      Cookies.remove("token");
      Cookies.remove("userId");
      setToken(null);
      setUserId(null);
    }
  };

  return (
    <AuthContext.Provider value={{ token, userId, setAuth }}>
      <ChatContext.Provider value={{ chatState, setChatState }}>
        {children}
      </ChatContext.Provider>
    </AuthContext.Provider>
  );
};

export default StateProvider;
