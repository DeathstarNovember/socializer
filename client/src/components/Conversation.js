import React, { useCallback } from "react";
import { useQuery } from "@apollo/react-hooks";
import gql from "graphql-tag";
import produce from "immer";
import { ErrorMessage, Loading, MessageThread, NewMessage } from "components";
import { Subscriber } from "containers";
import "./Conversation.css";

export const GET_CONVERSATION = gql`
  query GetConversation($id: String!) {
    conversation(id: $id) {
      id
      title
      messages {
        id
        body
        user {
          id
          name
          gravatarMd5
        }
      }
    }
  }
`;

export const MESSAGES_SUBSCRIPTION = gql`
  subscription onMessageCreated($conversationId: String!) {
    messageCreated(conversationId: $conversationId) {
      id
      body
      user {
        id
        name
        gravatarMd5
      }
    }
  }
`;

const Conversation = ({
  match: {
    params: { id },
  },
}) => {
  const { loading, error, data, subscribeToMore } = useQuery(GET_CONVERSATION, {
    variables: { id },
  });
  const subscribeToNew = useCallback(
    () =>
      subscribeToMore({
        document: MESSAGES_SUBSCRIPTION,
        variables: { conversationId: id },
        updateQuery: (prev, { subscriptionData }) => {
          if (!subscriptionData.data) return prev;
          const newMessage = subscriptionData.data.messageCreated;

          // Check that we don't already have the
          // message stored.
          if (
            prev.conversation.messages.find(
              (message) => message.id === newMessage.id,
            )
          ) {
            return prev;
          }

          return produce(prev, (next) => {
            next.conversation.messages.push(newMessage);
          });
        },
      }),
    [id],
  );

  if (loading) {
    return <Loading />;
  } else if (error) {
    return <ErrorMessage message={error.message} />;
  } else {
    return (
      <Subscriber subscribeToNew={subscribeToNew}>
        <div className="chat-layout d-flex flex-column pb-4">
          <div className="chat-content d-flex flex-column">
            <h5>{data.conversation.title}</h5>
            <hr />
            <MessageThread messages={data.conversation.messages} />
          </div>

          <NewMessage conversationId={id} />
        </div>
      </Subscriber>
    );
  }
};

export default Conversation;
