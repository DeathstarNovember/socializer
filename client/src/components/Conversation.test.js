import React from "react";
import { render, wait } from "@testing-library/react";
import { MockedProvider } from "@apollo/react-testing";
import { MemoryRouter } from "react-router-dom";
import { Subscriber } from "containers";
import { AuthContext } from "util/context";
import Conversation, {
  GET_CONVERSATION,
  MESSAGES_SUBSCRIPTION,
} from "./Conversation";

jest.mock("containers/Subscriber", () =>
  jest.fn().mockImplementation(({ children }) => children),
);

const getMock = {
  request: {
    query: GET_CONVERSATION,
    variables: { id: 1 },
  },
  result: {
    data: {
      conversation: {
        id: 1,
        title: "Jane Peters, Jack Smith",
        messages: [
          {
            id: 1,
            body: "Hi there",
            user: {
              id: 1,
              name: "Jane Peters",
              gravatarMd5: "abc",
            },
          },
          {
            id: 2,
            body: "Hello",
            user: {
              id: 2,
              name: "Jack Smith",
              gravatarMd5: "def",
            },
          },
        ],
      },
    },
  },
};

describe("Conversation", () => {
  it("renders correctly when loading", () => {
    const { container } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <MockedProvider mocks={[getMock]} addTypename={false}>
            <Conversation match={{ params: { id: 1 } }} />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    expect(container).toMatchSnapshot();
  });

  it("renders correctly when loaded", async () => {
    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <MockedProvider mocks={[getMock]} addTypename={false}>
            <Conversation match={{ params: { id: 1 } }} />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("Jane Peters, Jack Smith"));
    expect(container).toMatchSnapshot();
  });

  it("renders correctly after created message", async () => {
    Subscriber.mockImplementation((props) => {
      const { default: ActualSubscriber } = jest.requireActual(
        "containers/Subscriber",
      );
      return <ActualSubscriber {...props} />;
    });

    const mocks = [
      getMock,
      {
        request: {
          query: MESSAGES_SUBSCRIPTION,
          variables: { conversationId: 1 },
        },
        result: {
          data: {
            messageCreated: {
              id: 3,
              body: "How's it going?",
              user: {
                id: 1,
                name: "Jane Peters",
                gravatarMd5: "abc",
              },
            },
          },
        },
      },
    ];

    const { container, getByText } = render(
      <MemoryRouter>
        <AuthContext.Provider value={{ userId: 1 }}>
          <MockedProvider mocks={mocks} addTypename={false}>
            <Conversation match={{ params: { id: 1 } }} />
          </MockedProvider>
        </AuthContext.Provider>
      </MemoryRouter>,
    );
    await wait(() => getByText("How's it going?"));
    expect(container).toMatchSnapshot();
  });
});
