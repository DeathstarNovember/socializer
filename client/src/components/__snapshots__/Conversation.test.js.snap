// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Conversation renders correctly after created message 1`] = `
<div>
  <div
    class="chat-layout d-flex flex-column pb-4"
  >
    <div
      class="chat-content d-flex flex-column"
    >
      <h5>
        Jane Peters, Jack Smith
      </h5>
      <hr />
      <div
        class="chat-messages"
      >
        <div
          class="d-flex chat-others"
        >
          <div
            class="d-flex mb-1 chat-bubble-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle mt-1"
              height="30"
              src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
              width="30"
            />
            <div
              class="p-2 chat-bubble"
            >
              Hi there
            </div>
          </div>
        </div>
        <div
          class="d-flex chat-others"
        >
          <div
            class="d-flex mb-1 chat-bubble-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle mt-1"
              height="30"
              src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
              width="30"
            />
            <div
              class="p-2 chat-bubble"
            >
              Hello
            </div>
          </div>
        </div>
        <div
          class="d-flex chat-others"
        >
          <div
            class="d-flex mb-1 chat-bubble-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle mt-1"
              height="30"
              src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
              width="30"
            />
            <div
              class="p-2 chat-bubble"
            >
              How's it going?
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="new-message mt-2 card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="new-message"
        >
          <div
            class="form-group"
          >
            <input
              class="form-control"
              placeholder="What's on your mind?"
              rows="3"
              value=""
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Conversation renders correctly when loaded 1`] = `
<div>
  <div
    class="chat-layout d-flex flex-column pb-4"
  >
    <div
      class="chat-content d-flex flex-column"
    >
      <h5>
        Jane Peters, Jack Smith
      </h5>
      <hr />
      <div
        class="chat-messages"
      >
        <div
          class="d-flex chat-others"
        >
          <div
            class="d-flex mb-1 chat-bubble-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle mt-1"
              height="30"
              src="//www.gravatar.com/avatar/abc?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/abc?d=retro&r=g&s=60 2x"
              width="30"
            />
            <div
              class="p-2 chat-bubble"
            >
              Hi there
            </div>
          </div>
        </div>
        <div
          class="d-flex chat-others"
        >
          <div
            class="d-flex mb-1 chat-bubble-wrapper"
          >
            <img
              alt="Gravatar for undefined"
              class="react-gravatar rounded-circle mt-1"
              height="30"
              src="//www.gravatar.com/avatar/def?d=retro&r=g&s=30"
              srcset="//www.gravatar.com/avatar/def?d=retro&r=g&s=60 2x"
              width="30"
            />
            <div
              class="p-2 chat-bubble"
            >
              Hello
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="new-message mt-2 card"
    >
      <div
        class="card-body"
      >
        <form
          class=""
          data-testid="new-message"
        >
          <div
            class="form-group"
          >
            <input
              class="form-control"
              placeholder="What's on your mind?"
              rows="3"
              value=""
            />
          </div>
        </form>
      </div>
    </div>
  </div>
</div>
`;

exports[`Conversation renders correctly when loading 1`] = `
<div>
  <div
    class="d-flex justify-content-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      >
        Loading...
      </span>
    </div>
  </div>
</div>
`;
